poly-commit/src/hyrax/mod.rs

```diff
@@ -0,0 +1,533 @@
+/// [[WTsTW17]][hyrax].
+///
+/// [hyrax]: https://eprint.iacr.org/2017/1132.pdf
+///
+/// ### Future optimisations
+///
+/// - Add parallelisation. There is at least one natural place where
+///   parallelisation could bring performance gains: in essence, the prover
+///   commits to the polynomial by expressing it as an evaluation matrix and
+///   Pederson-multi-committing to each row. Each of this commitments can be
+///   computed independently from the rest, and therefore, in parallel. It is
+///   still to be seen how much of an improvement this would entail, since each
+///   Pederson multi-commitment boils down to a multi-exponentiation and this
+///   operation is itself parallelised.
+/// - Due to the homomorphic nature of Pedersen commitments, it is likely
+///   some of the following methods can be designed more efficiently than their
+///   default implementations: `batch_open`, `batch_check`,
+///   `open_combinations`, `check_combinations`. This is not discussed in the
+///   reference article, but the IPA and KZG modules might be a good starting
+///   point.
+/// - On a related note to the previous point, there might be a more
+///   efficient way to open several polynomials at a single point (this is the
+///   functionality of the `open` method) than the currently implemented
+///   technique, where only the computation of the vectors `L` and `R` is
+///   shared across polynomials.
+/// - The cited article proposes an optimisation in the section _Reducing the
+///   cost of proof-of-dot-prod_. It allows for non-square matrices (and hence
+///   removes the requirement for the number of variables to be even) and
+///   introduces a tradeoff between proof size and verifier time. It is
+///   probably worth pursuing.
+
+pub struct HyraxPC<
+    // The elliptic curve used for Pedersen commitments (only EC groups are
+    // supported as of now).
+    G: AffineRepr,
+    // A polynomial type representing multilinear polynomials
+    P: MultilinearExtension<G::ScalarField>,
+    // The sponge used in the protocol as random oracle
+    S: CryptographicSponge,
+> {
+    _phantom: PhantomData<(G, P, S)>,
+}
+
+impl<G, P, S> HyraxPC<G, P, S>
+where
+    G: AffineRepr,
+    P: MultilinearExtension<G::ScalarField>,
+    S: CryptographicSponge,
+{
+    /// Pedersen commitment to a vector of scalars as described in appendix A.1
+    /// of the reference article.
+    /// The caller must either directly pass hiding exponent `r` inside Some,
+    /// or provide an rng so that `r` can be sampled.
+    /// If there are `n` scalars, the first `n` elements of the key will be
+    /// multiplied by them in the same order, and its `n + 1`th element will be
+    /// multiplied by `r`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if both `r` and `rng` are None.
+    fn pedersen_commit(
+        key: &HyraxCommitterKey<G>,
+        scalars: &[G::ScalarField],
+        r: Option<G::ScalarField>,
+        rng: Option<&mut dyn RngCore>,
```

Comment:
Is this always a hiding commitment?

Reply:
I will let the expert talk! :D @Antonio95

Reply:
Hey there! So, in [its published form](https://eprint.iacr.org/2017/1132) (subsection 6.1), this PCS is always hiding indeed: it hides both the polynomial coefficients (by using an extra point in the Pedersen commitment key, as well as an additional random coefficient for each commitment) and the evaluation of the polynomial at the requested point, which is never learnt by the verifier.

One of the authors (J. Thaler) has mentioned a non-hiding version in private discussions (not involving me). I imagine this entails using non-hiding Pedersen commitments, but some more substantial changes would need to be made to the dot product argument (part of the opening and verification methods) as well. This has, as far as I know, never appeared in print.